libsupport/include/katana/Result.h

```diff
@@ -260,6 +260,40 @@ ResultErrno() {
   return std::error_code(errno, std::system_category());
 }
 
+namespace internal {
+
+template <class T>
+std::enable_if_t<!std::is_same<T, void>::value, T&&>
+extract_result_value(Result<T>&& result) {
+  return std::move(result.value());
+}
+
+inline int
+extract_result_value(Result<void>&&) {
+  return 0;
+}
+
+}  // namespace internal
+
+#define KATANA_CHECK_NAME(x, y) x##y
+
+#define KATANA_CHECK_IMPL(result_name, expression, ...)                        \
+  ({                                                                           \
+    auto result_name = (expression);                                           \
+    if (!result_name) {                                                        \
+      return result_name.error().WithContext(__VA_ARGS__);                     \
+    }                                                                          \
+    std::move(                                                                 \
+        ::katana::internal::extract_result_value(std::move(result_name)));     \
+  })
+
+#define KATANA_CHECK_CONTEXT(expression, ...)                                  \
+  KATANA_CHECK_IMPL(                                                           \
+      KATANA_CHECK_NAME(_error_or_value, __COUNTER__), expression,             \
+      __VA_ARGS__)
+
+#define KATANA_CHECK(expression) KATANA_CHECK_CONTEXT(expression, "backtrace")
```

Comment:
Does this make <Result>`.error()` obsolete? That is, we always return <Result>`.error().WithContext()`? I'm fine with that. Just wanted to confirm.

Reply:
`.error()` is still useful in cases where the current call is just an internal wrapper that doesn't add any useful context.

Reply:
But then `KATANA_CHECK` doesn't support or expose that right? My point is we either support both or support only `WithContext` - whether we use the macro or not.